Add a tutorial on how to create Build Scripts for a Platform. (#222)
MicroEJ/docs: Tutorials/tutorialCreateBuildScripts.rst

> Create Platform Build Scripts

Reviewer: Please add an anchor so we can reference the tuto elsewhere (e.g. `.. _tutorialCreateBuildScripts:`)

Author: Done

=============================

This tutorial describes all the steps to create platform build scripts and shows how to use them.

Intended Audience
-----------------

> The audience for this document is Platform engineers who want to

Reviewer: ```suggestion
The audience for this document are Platform engineers who want to
```

Author: Not that clear. I've picked this sentence from another tutorial. The subject is "the audience". https://ell.stackexchange.com/questions/77797/most-of-the-audience-is-are

Reviewer: You're correct, and I now remember I've been bitten by this one on the other tutorial :P

- validate their platform using MicroEJ test suites
- configure a full :ref:`BSP connection<bsp_connection>` with their platform

> - ease application development by removing the link to the BSP

Reviewer: What do you mean by "removing the link to the BSP"?

Author: I mean that the app developer does not need to be aware of the BSP, he only configures is platform.

Reviewer 2: 2 remarks to be discussed:
- build script**s** is not correct: rather use BSP scripts (composed of build script https://docs.microej.com/en/latest/PlatformDeveloperGuide/platformCreation.html#build-script-file and run script https://docs.microej.com/en/latest/PlatformDeveloperGuide/platformCreation.html#run-script-file)
- I think this tuto should be renamed "tutoConfigurePlatformBSPConnection" (Configure MicroEJ Platform BSP Connection)

Reviewer: We agreed on "build and run script".  @bguedas please you this naming throughout the tutorial.

Author: The introduction does not talk about BSP connection, but BSP connection is listed in 'intended audience' because someone who wants to configure a full BSP connection will find useful information in this tutorial too. This part only lists all the cases where reading the tutorial may be useful.

What do you think of someting like
```
The audience for this document is Platform engineers who want to
- write build and run scripts for their platform 
- validate their platform using MicroEJ test suites

Platform engineers who want to configure a full :ref:`BSP connection<bsp_connection>` with their platform will also find useful information in this tutorial.
```

Author: I've changed `build script` with `build and run scripts`

Prerequisites
-------------

This tutorial is a direct continuation of :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>` tutorial. It should have been completed before starting this one.

Introduction
------------

Build scripts are normalized entry points to

- build an application against a platform
- deploy and run the built application on a device.

External tools only need to run these scripts without additional knowledge about the toolchain or deployment tools.

See :ref:`Build Script File and Run Script File<bsp_connection_build_script>` subsections from MicroEJ Platform Creation guide for more information about these scripts. Script examples are provided in `Platform Qualification Tools <https://github.com/MicroEJ/PlatformQualificationTools/tree/master/framework/platform/scripts>`_ repository.

Overview
--------

> In  :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>`, the :ref:`BSP connection<bsp_connection>` is partial: the final binary is produced by invoking ``make`` in the FreeRTOS BSP.

Reviewer: That's incorrect, the connection is partial means that the property `root.dir` is defined in the application launcher instead of being defined in the platform.

Author: This part has been rewritten.

We use WSL to build the firmware and launch it with QEMU. We thus setup build scripts for these tasks: ``build.sh`` will build ``application.out`` whereas ``run.sh`` will launch it in QEMU.

The next sections will

- describe step-by-step how to ceate the build script both for shell and batch terminals. These scripts automate platform build and firmware execution in QUEMU as presented in :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>` tutorial.
- show a pratical usage of these buid scipts in a MicroEJ development flow. This will allow to configure Eclipse build the full firmware when building an application.

Create Build Scripts
--------------------

This section describes how to write build scripts.

> There are four scripts:

Reviewer: You present 2 scripts (build.xxx and run.xxx) not 4.

Author: Fixed

#. ``run.xxx`` which deploys and runs ``application.out`` on the device. In this tutorial, it will only run the application with QEMU instead of flashing it on real hardware.

Create a ``microej/scripts`` directory in BSP project:

.. code-block:: shell

  $ pwd
  /mnt/c/Users/user/src/tuto-from-scratch/FreeRTOS/FreeRTOS/Demo/CORTEX_LM3S6965_GCC
  $ mkdir microej/scripts

All the script created in the next sections will be put in this directory.

Create `build.sh` and `run.sh` Scripts
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. warning::

  Make sure the build scripts have the execution permission.

1. Create a file called ``build.sh`` in the ``microej/scripts`` directory with the following content:

  .. code-block:: shell

    #!/bin/bash

    # Save application current directory and jump one level above scripts
    CURRENT_DIRECTORY=$(pwd)
 
    # Move to the directory where the Makefile is
    cd $(dirname "$0")/../..

    # Build the firmware
    make

    # Copy output the the current directory while renaming it
    cp gcc/RTOSDemo.bin  $CURRENT_DIRECTORY/application.out

    # Restore application directory
    cd $CURRENT_DIRECTORY/

> 2. Verify that the scrit succesfully build our firmware and put it in the current directory with the name ``application.out``.

Reviewer: ```suggestion
2. Verify that the script successfully built our firmware and put it in the current directory with the name ``application.out``.
```

  .. code-block:: shell

    $ pwd
    /mnt/c/Users/user/src/tuto-from-scratch/FreeRTOS/FreeRTOS/Demo/CORTEX_LM3S6965_GCC
    $ make clean
    $ microej/scripts/build.sh
      CC    init/startup.c
      CC    main.c
      CC    ../../Source/list.c
      CC    ../../Source/queue.c
      CC    ../../Source/tasks.c
    [..]
      130 |  __attribute__( ( always_inline ) ) static inline uint8_t ucPortCountLeadingZeros( uint32_t ulBitmap )
          |                                                           ^~~~~~~~~~~~~~~~~~~~~~~
      LD    gcc/RTOSDemo.axf
    $ ls *.out
    application.out

3. Check that ``application.out`` successfully runs with QEMU:

  .. code-block:: shell

    $ qemu-system-arm -M lm3s6965evb -nographic -kernel application.out
    Hello, World! printf function is working.
    Hello World!
    QEMU: Terminated // press Ctrl-a x to end the QEMU session

> 4. Create a file called ``run.sh`` in the ``microej/scripts`` directory with the following content:

Reviewer: The run.xxx script should program the application.out located in the directory of execution.  In this case it is not necessary to copy it, but it might be worth mentioning it.

Reviewer: If possible, find a way to terminate the QEMU session

Author: I've tried, but didn't find a satisfying solution.

Reviewer: Ok,  Let's drop this for now then.

  .. code-block:: shell

    #!/bin/bash

    # Add some text to the console before launch
    echo -e "\033[0;32m## Start application in QEMU."
    echo -e "## Use 'Ctrl-a x' to quit.\e[0m"

    # Launch application with QEMU
    qemu-system-arm -M lm3s6965evb -nographic -kernel application.out

5. We can now run the application we just built with the ``run.sh`` script:

  .. code-block:: shell

    $ pwd
    /mnt/c/Users/user/src/tuto-from-scratch/FreeRTOS/FreeRTOS/Demo/CORTEX_LM3S6965_GCC
    $ microej/scripts/run.sh
    ## Start application in QEMU.
    ## Use 'Ctrl-a x' to quit.
    Hello, World! printf function is working.
    Hello World!

.. note::

>   This script is very simple because our application is just run with QEMU instead of a real hardware. To deploy the applcation on a device, the script would have to setup and call a flash tool. See for instance the build script for `Espressif-ESP-WROVER-KIT-V4.1 <https://github.com/MicroEJ/Platform-Espressif-ESP-WROVER-KIT-V4.1/blob/1.7.0/ESP32-WROVER-Xtensa-FreeRTOS-bsp/Projects/microej/scripts/build.sh>`_. 

Reviewer: ```suggestion
  This script is very simple because our application is just run with QEMU instead of a real hardware. To deploy the application on a device, the script would have to setup and call a flash tool. See for instance the build script for `Espressif-ESP-WROVER-KIT-V4.1 <https://github.com/MicroEJ/Platform-Espressif-ESP-WROVER-KIT-V4.1/blob/1.7.0/ESP32-WROVER-Xtensa-FreeRTOS-bsp/Projects/microej/scripts/build.sh>`_. 
```

Create `build.bat` and `run.bat` Scripts
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> As our toolchain has only be configured for Linux in WSL, we simply create wrappers that call shell scripts through WSL.

Reviewer: ```suggestion
As our toolchain has only be configured for Linux in WSL, we create wrappers that call shell scripts through WSL.
```

    The MicroEJ platform library (microejruntime.a) has been deployed to: 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\lib'.
    The MicroEJ platform header files (*.h) have been deployed to: 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\inc'.
    Execution of script 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\scripts\build.bat' started...
    LD    gcc/RTOSDemo.axf
    Current DIR /mnt/c/Users/user/Workspaces/_test_fw_tuto/HelloWorld/com.mycompany.Main
    Execution of script 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\scripts\build.bat' done.
    =============== [ Completed Successfully ] ===============

    SUCCESS

Reading the traces, we see that the `HelloWorld` application and the platform MicroEJ Platform library have been deployed to the BSP and the ```build.bat`` script has been called to rebuild the BSP. The output is then ``application.out`` binary that can be flashed on the board (or run on QEMU).

> Links the Platform to the BSP

Reviewer: ```suggestion
Link the Platform to the BSP
```

Reviewer: Not relevant anymore

    The MicroEJ platform library (microejruntime.a) has been deployed to: 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\lib'.
    The MicroEJ platform header files (*.h) have been deployed to: 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\inc'.
    Execution of script 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\scripts\build.bat' started...
      LD    gcc/RTOSDemo.axf
    Current DIR /mnt/c/Users/user/src/tuto-from-scratch/FreeRTOS/FreeRTOS/Demo/CORTEX_LM3S6965_GCC/microej/scripts
    Execution of script 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GC  C\microej\scripts\build.bat' done.
    =============== [ Completed Successfully ] ===============

    SUCCESS

  .. note::

    You can notice the difference in the second line of the trace that now says that the connection is `using platform option 'root.dir' in 'bsp/bsp.properties'.` instead of `using platform option 'deploy.bsp.root.dir'.` in the previous launch.

>  The application launcher does not know anymore where the BSP is located Nevertheless it still builds a full firmware ready to be flashed. The link to the BSP is now configured in the platform and every MicroEJ mono-sandboxed application projects can use it without extra configuration.

Reviewer: ```suggestion
 The application launcher does not know anymore where the BSP is located. Nevertheless it still builds a full firmware ready to be flashed. The link to the BSP is now configured in the platform and every MicroEJ mono-sandboxed application projects can use it without extra configuration.
```

Intended Audience
-----------------

The audience for this document is Platform engineers who want to

- validate their platform using MicroEJ test suites
- configure a full :ref:`BSP connection<bsp_connection>` with their platform
- ease application development by removing the link to the BSP

Prerequisites
-------------

This tutorial is a direct continuation of :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>` tutorial. It should have been completed before starting this one.

Introduction
------------

Build scripts are normalized entry points to

- build an application against a platform
- deploy and run the built application on a device.

External tools only need to run these scripts without additional knowledge about the toolchain or deployment tools.

See :ref:`Build Script File and Run Script File<bsp_connection_build_script>` subsections from MicroEJ Platform Creation guide for more information about these scripts. Script examples are provided in `Platform Qualification Tools <https://github.com/MicroEJ/PlatformQualificationTools/tree/master/framework/platform/scripts>`_ repository.

Overview
--------

> In  :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>`, the :ref:`BSP connection<bsp_connection>` the final binary is produced by invoking ``make`` in the FreeRTOS BSP. The command to type is dependant of the toolchain used. The application is then launched in QEMU but could have been instead flashed to a device with another specific command. This tutorial explain how to write `build` and `run` scripts for these two tasks. 

Reviewer: ```suggestion
In  :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>`, the final binary is produced by invoking ``make`` in the FreeRTOS BSP. The command to type is dependant of the toolchain used. The application is then launched in QEMU but could have been instead flashed to a device with another specific command. This tutorial explain how to write `build` and `run` scripts for these two tasks. 
```

The next sections will

> - describe step-by-step how to ceate the build and run scripts both for shell and batch terminals. These scripts automate platform build and firmware execution in QUEMU as presented in :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>` tutorial.

Reviewer 3: "QUEMU" ==> QEMU ?

> - describe step-by-step how to ceate the build and run scripts both for shell and batch terminals. These scripts automate platform build and firmware execution in QUEMU as presented in :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>` tutorial.

Reviewer: ```suggestion
- describe step-by-step how to ceate the build and run scripts both for bash and Windows Batch Script files. These scripts automate platform build and firmware execution in QUEMU as presented in :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>` tutorial.
```

      LD    gcc/RTOSDemo.axf
    $ ls *.out
    application.out

3. Check that ``application.out`` successfully runs with QEMU:

  .. code-block:: shell

    $ qemu-system-arm -M lm3s6965evb -nographic -kernel application.out
    Hello, World! printf function is working.
    Hello World!
    QEMU: Terminated // press Ctrl-a x to end the QEMU session

4. Create a file called ``run.sh`` in the ``microej/scripts`` directory with the following content:

  .. code-block:: shell

    #!/bin/bash

    # Add some text to the console before launch
    echo -e "\033[0;32m## Start application in QEMU."
    echo -e "## Use 'Ctrl-a x' to quit.\e[0m"

    # Launch application with QEMU
    qemu-system-arm -M lm3s6965evb -nographic -kernel application.out

5. We can now run the application we just built with the ``run.sh`` script:

  .. code-block:: shell

    $ pwd
    /mnt/c/Users/user/src/tuto-from-scratch/FreeRTOS/FreeRTOS/Demo/CORTEX_LM3S6965_GCC
    $ microej/scripts/run.sh
    ## Start application in QEMU.
    ## Use 'Ctrl-a x' to quit.
    Hello, World! printf function is working.
    Hello World!

.. note::

>   This script is very simple because our application is just run with QEMU instead of a real hardware. To deploy the application on a device, the script would have to setup and call a flash tool. See for instance the build and run scripts for `Espressif-ESP-WROVER-KIT-V4.1 <https://github.com/MicroEJ/Platform-Espressif-ESP-WROVER-KIT-V4.1/blob/1.7.0/ESP32-WROVER-Xtensa-FreeRTOS-bsp/Projects/microej/scripts/build.sh>`_. 

Reviewer 3: "instead of a real hardware" ==> should be "instead of real hardware" (or "instead of a real hardware board/device/something") I believe

Reviewer 2: fixed

Create `build.bat` and `run.bat` Scripts
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As our toolchain has only be configured for Linux in WSL, we create wrappers that call shell scripts through WSL.

> 1. Create a file called `build.bat` in the `microej/scripts` directory with the following content:

Reviewer: ```suggestion
1. Create a file called ``build.bat`` in the ``microej/scripts`` directory with the following content:
```

    [INFO ] Launching in Evaluation mode. Your UID is 0120202834374C4A.
    =============== [ Launching SOAR ] ===============
    =============== [ Launching Link ] ===============
    =============== [ Deployment ] ===============
    MicroEJ files for the 3rd-party BSP project are generated to 'C:\Users\user\src\tuto-from-scratch\workspace\HelloWorld\com.mycompany.Main\platform'.

    FAIL
    The following error occurred while executing this line:
    C:\Users\user\src\tuto-from-scratch\workspace\lm3s811evb-Platform-CM0_GCC48-0.0.1\source\scripts\deploy.xml:30: The following error occurred while executing this line:
    C:\Users\user\src\tuto-from-scratch\workspace\lm3s811evb-Platform-CM0_GCC48-0.0.1\source\scripts\deployInBSP.xml:97: The following error occurred while executing this line:
    C:\Users\user\src\tuto-from-scratch\workspace\lm3s811evb-Platform-CM0_GCC48-0.0.1\source\scripts\deployInBSP.xml:260: Option 'deploy.bsp.microejscript' is enabled but this Platform does no define a well-known location. Either update the Platform configuration (option 'deploy.bsp.microejscript.relative.dir' in 'bsp/bsp.properties') or disable this option.

> 7. Edit the file ``bsp/bsp.properties`` as follow:

Reviewer 3: I think it could be nice to remind to rebuild the platform after editing `bsp.properties`.

  .. code-block:: properties

    # Specify BSP external scripts files ('build.bat' and 'run.bat') parent directory.
    # This is a '/' separated directory relative to 'bsp.root.dir'.
    microejscript.relative.dir=microej/scripts

> 8. Run the `HelloWorld` application once again. This should print the following result:

Reviewer: ```suggestion
8. Build the `HelloWorld` application once again. This should print the following result:
```

    The MicroEJ platform library (microejruntime.a) has been deployed to: 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\lib'.
    The MicroEJ platform header files (*.h) have been deployed to: 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\inc'.
    Execution of script 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\scripts\build.bat' started...
    LD    gcc/RTOSDemo.axf
    Current DIR /mnt/c/Users/user/Workspaces/_test_fw_tuto/HelloWorld/com.mycompany.Main
    Execution of script 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S6965_GCC\microej\scripts\build.bat' done.
    =============== [ Completed Successfully ] ===============

    SUCCESS

Reading the traces, we see that the `HelloWorld` application and the platform MicroEJ Platform library have been deployed to the BSP and the ```build.bat`` script has been called to rebuild the BSP. The output is then ``application.out`` binary that can be flashed on the board (or run on QEMU).

> Finish full BSP connection (optional)

Reviewer: ```suggestion
Convert from partial BSP connection to full BSP connection (optional)
```

Intended Audience
-----------------

The audience for this document is Platform engineers who want to

- validate their platform using MicroEJ test suites
- configure a full :ref:`BSP connection<bsp_connection>` with their platform
- ease application development by removing the link to the BSP

Prerequisites
-------------

This tutorial is a direct continuation of :ref:`Create a MicroEJ Firmware From Scratch<create_firmware_from_scratch>` tutorial. It should have been completed before starting this one.

Introduction
------------

Build and run scripts are normalized entry points to

- build an application against a platform
- deploy and run the built application on a device.

External tools only need to run these scripts without additional knowledge about the toolchain or deployment tools.

See :ref:`Build Script File and Run Script File<bsp_connection_build_script>` subsections from MicroEJ Platform Creation guide for more information about these scripts. Script examples are provided in `Platform Qualification Tools <https://github.com/MicroEJ/PlatformQualificationTools/tree/master/framework/platform/scripts>`_ repository.

Overview
--------

> <<<<<<< HEAD

Reviewer: Looks like a merge leftover.  Please fix it.

Author: fixed
